Testscripts/Linux/ntttcp_monitor_throughput.sh

```diff
@@ -0,0 +1,167 @@
+        # waiting for ntttcp process to start and get a stable throughput
+        sleep $((${monitor_interval} + 10))
+
+        # start monitoring on client
+        while [ $end_time -gt $(date +%s) ]
+        do
+            date_back_interval=$(date -d "-${monitor_interval} seconds" +"%Y-%m-%d %H:%M:%S")
+            # TODO: should we use tail, or direct use ./monitor_ntttcp.log? need more testing to select the most efficient
+            # ntttcp log real-time throughput every 500ms, so interval
+            tail -$((2 * ${monitor_interval} + 5)) "${log_dir}"/monitor_ntttcp.log > "${log_dir}"/tail_throughput.log
+            sed -un '/'"${date_back_interval}"'/,$ s/.*/&/w '"${log_dir}"'/date_back_interval.log' "${log_dir}"/tail_throughput.log
+            # if ./date_back_interval.log is empty, means ntttcp aborted due to networking device broken
+            # exit and break from while loop
+            if [[ ! -s "${log_dir}/date_back_interval.log" ]]; then
+                break
+            fi
+            # also if "${log_dir}"/date_back_interval.log has some more columns that's not 'throughput|blank_space|unit' format
+            # means networking device broken or ntttcp running down, exit and break from while loop
+            # due to performance consideration, merge into above break for ./date_back_interval.log is empty
+            sed -uni 's/.*Real-time\sthroughput:\s\([0-9.]*\)\(.*bps\)/\1 \2/p' "${log_dir}"/date_back_interval.log
+
+            # average_throughput|blank_space|unit
+            now_state=($(awk 'BEGIN{sum = 0} {sum += $1; now_unit = $2 } END{average = sum / NR; print average,now_unit}' "${log_dir}"/date_back_interval.log))
+
+            if [[ (1 = $(echo "${now_state[0]} != 0" | bc)) && (-n "${now_state[1]}") ]]; then
+                if [[ -z ${initialized} ]]; then
+                    unit_baseline=${now_state[1]}
+                    throughput_baseline=${now_state[0]}
+                    throughput_max=${throughput_baseline}
+                    throughput_min=${throughput_baseline}
+                    threshold_upper=$(awk -v tol="${tolerance}" -v thr="${throughput_baseline}" "BEGIN {print (1 + tol) * thr}")
+                    threshold_lower=$(awk -v tol="${tolerance}" -v thr="${throughput_baseline}" "BEGIN {print (1 - tol) * thr}")
+                    initialized="initialized"
+                elif [[ (1 = $(echo "${now_state[0]} > ${threshold_upper} ||  ${now_state[0]} < ${threshold_lower}" | bc)) || (${now_state[1]} != ${unit_baseline}) ]]; then
+                    mv "${log_dir}"/date_back_interval.log "${log_dir}"/out_of_tolerance@${now_state[0]}_$(date +"%Y_%m_%d_%H_%M_%S")
+                fi
+                echo "TestRunning, EndTime: ${end_time_readable}, Baseline: ${throughput_baseline} ${unit_baseline},  Tolerance: +/- ${tolerance}, Now throughput: ${now_state[0]} ${now_state[1]}" > "${log_dir}"/state.txt
```

Comment:
Should we use some different file than State.txt. As this filename is well known for LISAv2 framework to indicate Test exit state. Just concerned about user experience here.

Reply:
This long-term running test case, we don't use that to indicate test FAIL/COMPLETED,  we use 'state.txt' to save the current 'state' of test running or not, and the throughput.  The file name just matching the purpose. So use it.